sdk/workflow-tengo/src/pt/util.lib.tengo

```diff
@@ -0,0 +1,173 @@
+                    index < len(params.axes),
+                    "Column parent (%d) must be strictly less than the number of axes (%d)",
+                    index,
+                    len(params.axes)
+                )
+            }
+        }
+    }
+
+    // Check that partitionKeyLength is strictly less than the number of axes
+    ll.assert(
+        params.partitionKeyLength == undefined || params.partitionKeyLength < len(params.axes),
+        "partitionKeyLength (%d) must be strictly less than the number of axes (%d)",
+        params.partitionKeyLength,
+        len(params.axes)
+    )
+}
+
+unique := func(list) {
+    uniqueMap := {}
+    for _, item in list {
+        uniqueMap[item] = item
+    }
+    uniqueList := []
+    for _, item in uniqueMap {
+        uniqueList = append(uniqueList, item)
+    }
+    return uniqueList
+}
+
+makeFrameSpecs := func(params) {
+	axes := []
+    for entry in params.axes {
+        spec := {
+            type: entry.spec.type,
+            name: entry.spec.name
+        }
+        if !is_undefined(entry.spec.domain) && len(entry.spec.domain) > 0 {
+            spec.domain = entry.spec.domain
+        }
+        if !is_undefined(entry.spec.annotations) && len(entry.spec.annotations) > 0 {
+            spec.annotations = entry.spec.annotations
+        }
+        if !is_undefined(entry.spec.parentAxes) && len(entry.spec.parentAxes) > 0 {
+            spec.parentAxes = unique(entry.spec.parentAxes)
+        }
+        axes = append(axes, spec)
+    }
+    columns := {}
+    for entry in params.columns {
+        spec := {
+            kind: constants.KIND_P_COLUMN,
+            axesSpec: axes,
+            valueType: entry.spec.valueType,
+            name: entry.spec.name
+        }
+        if !is_undefined(entry.spec.domain) && len(entry.spec.domain) > 0 {
+            spec.domain = entry.spec.domain
+        }
+        if !is_undefined(entry.spec.annotations) && len(entry.spec.annotations) > 0 {
+            spec.annotations = entry.spec.annotations
+        }
+        if !is_undefined(entry.spec.parentAxes) && len(entry.spec.parentAxes) > 0 {
+            spec.parentAxes = unique(entry.spec.parentAxes)
+        }
+        columns[entry.column] = spec
+    }
+    return columns
+}
+
+makeWriteFrameStep := func(frameName, params) {
+	step := {
+        type: "write_frame",
+        name: frameName,
+        axes: [],
+        columns: [],
+        partitionKeyLength: 0
+    }
+    for entry in params.axes {
+        axis := {
+            column: entry.column,
+            type: entry.spec.type
+        }
+        step.axes = append(step.axes, axis)
+    }
+    for entry in params.columns {
+        column := {
+            column: entry.column,
+            type: entry.spec.valueType
+        }
+        step.columns = append(step.columns, column)
+    }
+    if !is_undefined(params.partitionKeyLength) {
+        step.partitionKeyLength = params.partitionKeyLength
+    }
+    return step
+}
```

Comment:
There are helper functions that can be used here:
- https://github.com/milaboratory/platforma/blob/a6014a195b35277204d601c74a7536e6f88fbe54/sdk/workflow-tengo/src/slices.lib.tengo#L123-L130
- https://github.com/milaboratory/platforma/blob/a6014a195b35277204d601c74a7536e6f88fbe54/sdk/workflow-tengo/src/maps.lib.tengo#L159-L197